### DISCUSSION TITLE
Inquiries about the http/3 and crowdsec rule

### DISCUSSION THREAD

**Discussion by dalso0418** [ZoeyVid/NPMplus#479] [Q&A] [answered]
Hello. You're asking a lot of questions!

It's because the project is so good, so please understand.

I have a few more questions.

I currently opened all 443 ports with TCP/UDP, but http3check.When checked on the net, it seems that communication is not possible through http3. http 1.1

![image](https://github.com/ZoeyVid/NPMplus/assets/107911055/c10b3b18-fe1b-4386-94bf-bec2a78e4cfc)
![image](https://github.com/ZoeyVid/NPMplus/assets/107911055/c77c2861-bbc6-4694-b48e-00eb603fc918)

I have another question about Crowdsec.
Currently, I can only see the list as below in the basic drawsec rule, but I would like to ask if there is a way to test if things detected in nginx are blocked.

cscli collections list

![image](https://github.com/ZoeyVid/NPMplus/assets/107911055/eb831339-8053-4549-82ea-ad3d6de2d3a7)

Thank you.
*This content was written through a translator, so please understand even if there are mistranslations.

**Comment by Zoey2936** [comment]
https://http3check.net/?host=zoeyvid.de returns HTTP/3 for me, so did you exposed udp  in all firewalls?

For crowdsec testing:
docker exec crowdsec cscli decisions add --ip 1.2.3.4 --duration 1m
Use your ip and open the site

**Comment by dalso0418** [comment]
Thank you for your reply.
but, I am using an unifi router and tcp/udp is open on both 80/443.

![image](https://github.com/ZoeyVid/NPMplus/assets/107911055/efacf1a5-5d81-4862-8f04-3990b8014cd2)

**Reply by Zoey2936** [reply]
can you maybe say your domain, so i can check?

**Reply by dalso0418** [reply]
test.dalso.org Thank you for checking.

**Reply by Zoey2936** [reply]
can you plase run `grep -rn /opt/npm/nginx -e "reuse"`

sometimes it works:
```console
curl --http3 https://test.dalso.org -I
HTTP/3 200 
date: Fri, 10 Nov 2023 12:25:11 GMT
content-type: text/html
content-length: 612
vary: accept-encoding
last-modified: Tue, 08 Nov 2022 06:24:37 GMT
etag: "6369f625-264"
accept-ranges: bytes
alt-svc: h3=':443'; ma=86400
x-xss-protection: 0
x-frame-options: SAMEORIGIN
x-content-type-options: nosniff
referrer-policy: strict-origin-when-cross-origin
content-security-policy: upgrade-insecure-requests
strict-transport-security: max-age=31536000; includeSubDomains; preload
```

**Reply by dalso0418** [reply]
![image](https://github.com/ZoeyVid/NPMplus/assets/107911055/75a733a4-ed7e-4664-bb17-8b41461b28ac)
It's shown above.

**Reply by Zoey2936** [reply]
Please run: `docker run --rm zoeyvid/curl-quic --http3-only -IL https://domain` and `docker run --rm zoeyvid/curl-quic --http3-only -Ik https://internal-ip`

**Reply by dalso0418** [reply]
When you test with domain, you get infinite loading. Why is that

docker run --rm zoeyvid/curl-quic --http3 -IL https://test.dalso.org  
% Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:06:21 --:--:--     0^C

```
# docker run --rm zoeyvid/curl-quic --http3 -Ik http://192.168.1.60:83
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0   612    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
HTTP/1.1 200 OK
Server: nginx/1.18.0 (Ubuntu)
Date: Sat, 11 Nov 2023 01:59:42 GMT
Content-Type: text/html
Content-Length: 612
Last-Modified: Tue, 08 Nov 2022 06:24:37 GMT
Connection: keep-alive
ETag: "6369f625-264"
Accept-Ranges: bytes
```

**Reply by Zoey2936** [reply]
Please test the ip with https

**Reply by dalso0418** [reply]
For internal ip, http is in use.

client -> NPMPLUS(http3) -> http reverse proxy
It's structured.

**Reply by Zoey2936** [reply]
So you have a second proxy in front of npmplus?

**Reply by Zoey2936** [reply]
Please send me this file: /opt/npm/nginx/default.conf

**Reply by dalso0418** [reply]
No second proxy!
80/443 (tcp/udp) is ported to npmplus in UDM.

It's about the route you mentioned, but the file is empty.
(I updated it to the latest version today)
![image](https://github.com/ZoeyVid/NPMplus/assets/107911055/f7e714ca-f6f4-4b4d-8a39-99eafc0fa898)

**Reply by Zoey2936** [reply]
The file is empty? Then there is the problem please recreate it using the webui settings->default site

**Reply by Zoey2936** [reply]
please update to the latest version, should be fixed now

**Reply by dalso0418** [reply]
As you said, default.conf was created normally.

However, http3 is still not applied to services delivered to reverse proxy

![image](https://github.com/ZoeyVid/NPMplus/assets/107911055/d963a6e0-3582-4c95-8794-6fb69202a1c2)

**Reply by Zoey2936** [reply]
works:
```
root@PC-Zoey:~# curl --http3 -IL https://test.dalso.org
HTTP/3 200
date: Sun, 12 Nov 2023 12:28:02 GMT
content-type: text/html
content-length: 612
vary: accept-encoding
last-modified: Tue, 08 Nov 2022 06:24:37 GMT
etag: "6369f625-264"
accept-ranges: bytes
alt-svc: h3=':443'; ma=86400
x-xss-protection: 0
x-frame-options: SAMEORIGIN
x-content-type-options: nosniff
referrer-policy: strict-origin-when-cross-origin
content-security-policy: upgrade-insecure-requests
strict-transport-security: max-age=31536000; includeSubDomains; preload
```

**Reply by dalso0418** [reply]
http3check.net
In addition, it is marked as not active on other sites.

https://http3test.seowebchecker.com/
![image](https://github.com/ZoeyVid/NPMplus/assets/107911055/121be0c3-2f3d-4f08-9135-d822c0c2b553)

On the other hand, in the case of other sites operating as caddy or your site is marked normally.https://http3check.net/?host=zoeyvid.de

![image](https://github.com/ZoeyVid/NPMplus/assets/107911055/fbf9e580-10b9-4d42-a1a5-7e83743d2e05)

**Reply by Zoey2936** [reply]
not sure, but testing with curl works....

**Reply by dalso0418** [reply]
Enough! Thank you for checking so far :)

**Reply by dalso0418** [reply]
I think I found a difference.
(Change settings It was recognized as normal in http3check for a while.)

The default.conf is set as below, but the vhost created by proxy_host is missing http3on, and the option is slightly different. " "Back."

**default.conf**
```
add_header Alt-Svc 'h3=":443"; ma=86400';
  http3 on;
```

**proxy host.conf**

`more_set_headers "Alt-Svc: h3=':443'; ma=86400";`

**Reply by Zoey2936** [reply]
http3 on; is set gloablly

**Reply by dalso0418** [reply]
I see, it was a strange experience.
I was wondering because it was recognized as http3 for a while while changing the settings (there was also a certificate error for a while)

Let me test a few more.

**Reply by dalso0418** [reply]
I think I found it.
I don't know where it was applied, but it seems to be applied through custom code on the Advanced side.
Please review it.

++There seems to be a problem with the automatically generated header ',"
Adding add_header alt-svc 'h3=":443"; ma=86400'; works fine.

![image](https://github.com/ZoeyVid/NPMplus/assets/107911055/05d27ad1-9427-483b-aa85-93e497a566ae)


![image](https://github.com/ZoeyVid/NPMplus/assets/107911055/6b1d71bc-d6a6-451a-90c6-f31e3064ce63)

**Reply by Zoey2936** [reply, marked as answer]
found the error, thanks for the debugging:
more_set_headers "Alt-Svc: h3=':443'; ma=86400";
need to be:
more_set_headers 'Alt-Svc: h3=":443"; ma=86400';

**Reply by Zoey2936** [reply]
will release a bug fix now

**Reply by dalso0418** [reply]
I'm glad to hear that it was helpful.

It's a fantastic project :)

**Reply by Zoey2936** [reply]
please report back if that fix helped

**Reply by dalso0418** [reply]
It's perfect. It's applied well.

**Comment by Zoey2936** [comment]
Did crowdsec at least worked?

**Reply by dalso0418** [reply]
Crowdsec blocking works well with the command you told me.

I was wondering if the cscli collections list above should include a list related to nginx if it is based on the log coming into the nginx proxy manager.

**Reply by Zoey2936** [reply]
collections has nothing todo with bouncers

**Comment by uaysk** [comment]
Hello, I looked up the official documentation related to nginx quic and found that the reuseport option was used instead of the quic option. 
https://nginx.org/en/docs/quic.html

```
http {
    log_format quic '$remote_addr - $remote_user [$time_local] '
                    '"$request" $status $body_bytes_sent '
                    '"$http_referer" "$http_user_agent" "$http3"';

    access_log logs/access.log quic;

    server {
        # for better compatibility it's recommended
        # to use the same port for quic and https
        listen 8443 quic reuseport; #The official documentation uses quic and reuseport
        listen 8443 ssl;

        ssl_certificate     certs/example.com.crt;
        ssl_certificate_key certs/example.com.key;

        location / {
            # required for browsers to direct them to quic port
            add_header Alt-Svc 'h3=":8443"; ma=86400';
        }
    }
}
```

It seems that only the quic option is used in the npm plus configuration file.

```
server {
  set $forward_scheme http;
  set $server         "localhost";
  set $port           81;

  listen unix:/run/nginx-14.sock;

  listen 0.0.0.0:80;
  listen [::]:80;


  listen 0.0.0.0:443 ssl;
  listen [::]:443 ssl;

  listen 0.0.0.0:443 quic;
  listen [::]:443 quic;

  more_set_headers "Alt-Svc: h3=':443'; ma=86400";
```

I'm not sure exactly what the reuseport option does. However, it seems to be used with quic in other documents besides the official document.

**Reply by Zoey2936** [reply]
reuseport is used inside the default.conf, it can and must only be used once

**Reply by uaysk** [reply]
Thank you for answer

**Reply by Zoey2936** [reply]
I've found the error, the default.conf is created empty, will be fixed in the next release

**Reply by uaysk** [reply]
Thank you, I updated docker image and it works fine
